src/engine/graphics/shaders.cpp

```diff
@@ -0,0 +1,210 @@
+#include "shaders.h"
+
+static std::map<std::string, hwShaderProgram*> programs;
+static hwShaderProgram* fallbackShaderProgram = nullptr;
+
+// For resetting following rebuild
+static std::string lastProgramName;
+
+/**
+ * Validate the default shader set has been loaded.
+ */
+static void Shaders_ValidateDefault() {
+	const char* defaultShaders[] = {
+		"generic_untextured",
+		"generic_textured",
+		"generic_textured_lit",
+	};
+
+	for ( unsigned int i = 0; i < plArrayElements( defaultShaders ); ++i ) {
+		hwShaderProgram* program = Shaders_GetProgram( defaultShaders[ i ] );
+		if ( program == nullptr ) {
+			Error( "Failed to fetch default shader, \"%s\"!\n", defaultShaders[ i ] );
+		}
+
+		if ( i == 0 ) {
+			fallbackShaderProgram = program;
+		}
+	}
+}
+
+static void Shaders_CacheShaderProgram( const char* path ) {
+	const char* fileName = plGetFileName( path );
+	if ( fileName == nullptr ) {
+		LogWarn( "Failed to get filename for shader program, \"%s\"!\n", path );
+		return;
+	}
+
+	// Provide a copy, minus the extension
+	std::string shortFileName( fileName, 0, strlen( fileName ) - 8 );
+
+	try {
+		ScriptConfig config( path );
+
+		config.EnterChildNode( "gl3" );
+
+		std::string vertPath = config.GetStringProperty( "vertPath" );
+		std::string fragPath = config.GetStringProperty( "fragPath" );
+
+		if ( vertPath.empty() || fragPath.empty()) {
+			// GetStringProperty throws warnings for this case, so just return
+			return;
+		}
+
+		hwShaderProgram* program = new hwShaderProgram( vertPath, fragPath );
+		programs.insert( std::pair<std::string, hwShaderProgram*>( shortFileName, program ));
+	} catch ( const std::exception& error ) {
+		LogWarn( "Failed to register shader program, %s!\n", error.what());
+	}
+}
+
+static void Shaders_ClearPrograms() {
+	for ( const auto& program : programs ) {
+		program.second->Disable();
+		delete program.second;
+	}
+
+	programs.clear();
+}
+
+static void Shaders_CachePrograms() {
+	Shaders_ClearPrograms();
+
+	char scanPath[PL_SYSTEM_MAX_PATH];
+	snprintf( scanPath, sizeof( scanPath ), "%s/shaders", u_get_base_path());
+	plScanDirectory( scanPath, "program", Shaders_CacheShaderProgram, false );
+
+	// Recursive searches through dependencies... This needs revisiting!
+	const char* modPath = u_get_mod_path();
+	if ( modPath[ 0 ] != '\0' ) {
+		snprintf( scanPath, sizeof( scanPath ), "%s/shaders", u_get_full_path());
+		plScanDirectory( scanPath, "program", Shaders_CacheShaderProgram, false );
+	}
+
+	Shaders_ValidateDefault();
+}
+
+static void Cmd_RebuildShaderPrograms( unsigned int argc, char* argv[] ) {
+	Shaders_CachePrograms();
+}
+
+static void Cmd_RebuildShaderProgram( unsigned int argc, char* argv[] ) {
+	u_unused( argc );
+	u_unused( argv );
+
+	try {
+		for ( const auto& program : programs ) {
+			program.second->Rebuild();
+		}
+	} catch ( ... ) {
+		Error( "Failed to rebuild shader programs, check logs for details!\n" );
+	}
+
+	Shaders_SetProgramByName( lastProgramName );
+}
```

Comment:
This function needs to be finished, was supposed to only rebuild one specified shader rather than rebuilding the lot.